# Hail Tutorial Notebooks

The notebooks in this directory show examples of running common Hail operations on the DNAnexus platform. Each notebook can be viewed individually as needed or in this general order:

- Import pVCF Genomic Data with Hail (_pVCF_import.ipynb_)
- Import BGEN Genomic Data with Hail (_BGEN_import.ipynb_)
- Filter Genomic Data with Hail: Chromosomes and Positions (_filter_chrpos.ipynb_)
- Filter Genomic Data with Hail: Sample IDs (_filter_sampleid.ipynb_)
- Filter Genomic Data with Hail: Variant IDs (_filter_varid.ipynb_)
- Replace Sample IDs with Hail (_replace_id.ipynb_)
- Variant Annotation with Hail: Variant Effect Predictor (VEP) (_MatrixTable_variant_annotation_with_VEP.ipynb_)
- Variant Annotation with Hail: Annotation Database (_MatrixTable_variant_annotation_with_db.ipynb_)
- pre-GWAS with Hail: Locus QC (_locus_qc.ipynb_)
- pre-GWAS with Hail: Sample QC (_sample_qc.ipynb_)
- GWAS with Hail (_gwas.ipynb_)
- Visualize GWAS results with Hail (_gwas_vis.ipynb_)
- Annotate GWAS results with Hail: Variant Effect Predictor (VEP) (_gwas_ann_vep.ipynb_)
- Annotate GWAS results with Hail: Annotation Database (_gwas_ann_haildb.ipynb_)
- Export Genomic Data as BGEN with Hail (_export_bgen.ipynb_)


See https://documentation.dnanexus.com/science/using-hail-to-analyze-genomic-data for additional information on each notebook, including JupyterLab with Spark Cluster app launch specs for different datasets.


## As-Is Software Disclaimer

This content in this repository is delivered “As-Is”. Notwithstanding anything to the contrary, DNAnexus will have no warranty, support, liability or other obligations with respect to Materials provided hereunder.