{"metadata": {"kernelspec": {"display_name": "Python 3 (ipykernel)", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.9.16"}}, "nbformat_minor": 5, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# \u201cdx extract_assay germline\u201d in Python\n<hr/>\n***As-Is Software Disclaimer***\n\nThis content in this repository is delivered \u201cAs-Is\u201d. Notwithstanding anything to the contrary, DNAnexus will have no warranty, support, liability or other obligations with respect to Materials provided hereunder.\n\n<hr/>\n\nThis notebook demonstrates usage of the dx command `extract_assay germline` to:\n* Get a list of sample-ids based corresponding to a set of RSIDs from a dataset.\n* Retrieve all the alleles in one gene and their associated annotation from a cohort.\n\n<a href=\"https://github.com/dnanexus/OpenBio/blob/master/LICENSE.md\">MIT License</a> applies to this notebook.", "metadata": {}, "id": "61b6d9d8"}, {"cell_type": "markdown", "source": "## Preparing your environment\n### Launch spec:\n\n* App name: JupyterLab with Python, R, Stata, ML\n* Kernel: Python\n* Instance type: mem1_ssd1_v2_x2\n* Cost: < $0.1\n* Runtime: < 2 min\n* Data description: Input for this notebook is a v3.0 Dataset or Cohort object ID. In the example considered below, the dataset and cohort have 50K samples and around 16M variants.", "metadata": {}, "id": "1ddd1c3c"}, {"cell_type": "markdown", "source": "### Install DNAnexus supported package, dxpy", "metadata": {}, "id": "26d26520"}, {"cell_type": "code", "source": "# For dx extract_assay germline, dxpy must be v0.349.1 or greater\n# However, a more recent version of dxpy on PyPI may already be available\n# and installed, making the below \"pip\" install unecessary.\n!pip3 install -U dxpy==0.363.0", "metadata": {}, "execution_count": null, "outputs": [], "id": "e03af894"}, {"cell_type": "code", "source": "import subprocess\nimport dxpy\nimport pandas as pd\nimport os\nimport glob\nimport json\npd.set_option('display.max_columns', None)", "metadata": {}, "execution_count": null, "outputs": [], "id": "7e0722a6"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"germline\", \"--help\"]\nhelp_text = subprocess.check_output(cmd)\nprint(help_text.decode(), end=\"\\n\")", "metadata": {}, "execution_count": null, "outputs": [], "id": "572ec4cd"}, {"cell_type": "markdown", "source": "### Assign environment variables", "metadata": {}, "id": "23e4126c"}, {"cell_type": "code", "source": "# The referenced dataset is not a public dataset and listed here only to demonstrate as an example input.\n# The user will need to supply a permissible and valid project ID and record ID\n\n# Assign a project qualified dataset, project-id:record-id\ndataset=\"project-GXYXjj00j49b5qX2Kzq1qbZk:record-FykXjyj0pZ7B8gKvKkxFF7QJ\"\ncohort=\"project-GXYXjj00j49b5qX2Kzq1qbZk:record-GXYkQyQ0j49jZ2fKpKZbxZ57\"", "metadata": {}, "execution_count": null, "outputs": [], "id": "c318e0a5"}, {"cell_type": "markdown", "source": "## Extract data from a dataset", "metadata": {}, "id": "5600aba9"}, {"cell_type": "markdown", "source": "### 1. Explore the genetic assays of a dataset\n\nCheck which genetic variant assays are available in the dataset using the `--list-assays` flag.", "metadata": {}, "id": "dc2eed6f"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"germline\", dataset, \"--list-assays\"]\nlist_assay = subprocess.check_output(cmd)\nprint(list_assay.decode(), end=\"\\n\")", "metadata": {}, "execution_count": null, "outputs": [], "id": "90e9b77d"}, {"cell_type": "markdown", "source": "### 2. Retrieve genomic data from the dataset", "metadata": {}, "id": "32b5e542"}, {"cell_type": "markdown", "source": "Data may be retrieved using one of three distinct options, `--retrieve-allele`, `--retrieve-annotation`, or `--retrieve-genotype`. Each option accepts a JSON object as input for filtering purposes, where the object is supplied either as a \\*.json file or as a string. For additional help, use `--json-help` trailing one of the `--retrieve-allele`, `--retrieve-annotation`, or `--retrieve-genotype` options for both a template and a list of filters available to the respective option.", "metadata": {}, "id": "9a66ae25"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"germline\", dataset, \"--retrieve-allele\", \"--json-help\"]\nallele_json_help = subprocess.check_output(cmd).decode()\nprint(allele_json_help)", "metadata": {}, "execution_count": null, "outputs": [], "id": "b1f4f4a3"}, {"cell_type": "markdown", "source": "#### Example on how to create a JSON of filters to retrieve data\n\nA common way to create a json object used for retrieving data is to update the json template from `--json-help` with the filters needed for the situation.", "metadata": {}, "id": "b91e7dbc"}, {"cell_type": "code", "source": "allele_filter_dict = {\n    \"rsid\":\n        [\"rs1342568097\", \n         \"rs1267100748\"], \n    \"type\":\n        [\"SNP\"]\n}\nallele_filter = json.dumps(allele_filter_dict)\n\nwith open(\"allele_filter.json\", \"w\") as op:\n    op.write(allele_filter)", "metadata": {}, "execution_count": null, "outputs": [], "id": "458af8b3"}, {"cell_type": "code", "source": "cmd = [\"cat\", \"allele_filter.json\"]\nallele_filter_file = subprocess.check_output(cmd).decode()\nprint(allele_filter_file)", "metadata": {}, "execution_count": null, "outputs": [], "id": "fc7c9051"}, {"cell_type": "markdown", "source": "#### Use the generated JSON file and retrieve allele data\n\nList allele IDs and related information based on the allele filters mentioned in the allele_filter.json and save the results in the `retrieve_allele_output.tsv` file using the `--output OUTPUT` option.", "metadata": {}, "id": "44dfb838"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"germline\", dataset, \"--retrieve-allele\", \"allele_filter.json\", \"-o\", \"retrieve_allele_output.tsv\"]\nsubprocess.check_call(cmd)", "metadata": {}, "execution_count": null, "outputs": [], "id": "63c4cf16"}, {"cell_type": "code", "source": "retrieve_allele_output = pd.read_csv(\"retrieve_allele_output.tsv\", sep=\"\\t\")\nretrieve_allele_output.head()", "metadata": {}, "execution_count": null, "outputs": [], "id": "6bc57769"}, {"cell_type": "markdown", "source": "#### Retrieve sample IDs based on allele IDs\n\nUse the allele IDs generated in the previous step as a filter and get sample IDs and related genotype information using the `--retrieve-genotype` filter. Here, we use a JSON string for filtering as opposed to a \\*.json file. ", "metadata": {}, "id": "d72694bd"}, {"cell_type": "code", "source": "# Create filter\ngenotype_filter = str({\"allele_id\": list(retrieve_allele_output[\"allele_id\"])})\n\n# Run extract_assay command\ncmd = [\"dx\", \"extract_assay\", \"germline\", dataset, \"--retrieve-genotype\", genotype_filter.replace(\"'\", '\"'), \"--output\", \"retrieve_genotype_output.tsv\"]\nsubprocess.check_call(cmd)", "metadata": {}, "execution_count": null, "outputs": [], "id": "51a8a8a3"}, {"cell_type": "code", "source": "retrieve_genotype_output = pd.read_csv(\"retrieve_genotype_output.tsv\", sep=\"\\t\")\nretrieve_genotype_output.head()", "metadata": {}, "execution_count": null, "outputs": [], "id": "1bdc8df0"}, {"cell_type": "markdown", "source": "Get a list of sample IDs", "metadata": {}, "id": "3db065d7"}, {"cell_type": "code", "source": "print(\"List of sample IDs with previously derived allele IDs.\")\nprint(list(retrieve_genotype_output[\"sample_id\"]))", "metadata": {}, "execution_count": null, "outputs": [], "id": "6fcb75ac"}, {"cell_type": "markdown", "source": "## Extract data from a cohort\n\nHere we will retrieve all the alleles in the gene `TP53` and their associated annotations from a saved cohort.", "metadata": {}, "id": "4441e076"}, {"cell_type": "markdown", "source": "#### Check the available filters in the `--retrieve-annotation` option.", "metadata": {}, "id": "9d4f6811"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"germline\", cohort, \"--retrieve-annotation\", \"--json-help\"]\nannotation_json_help = subprocess.check_output(cmd).decode()\nprint(annotation_json_help)", "metadata": {}, "execution_count": null, "outputs": [], "id": "f3aca5ec"}, {"cell_type": "markdown", "source": "#### We will use the `gene_name` filter and save the output data to a .tsv file.", "metadata": {}, "id": "d966530a"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"germline\", cohort, \"--retrieve-annotation\", \"{\\\"gene_name\\\": [\\\"TP53\\\"]}\", \"-o\", \"TP53_allele_annotation.tsv\"]\nsubprocess.check_call(cmd)", "metadata": {}, "execution_count": null, "outputs": [], "id": "91f13557"}, {"cell_type": "code", "source": "TP53_allele_annotation = pd.read_csv(\"TP53_allele_annotation.tsv\", sep=\"\\t\")\nTP53_allele_annotation.head()", "metadata": {}, "execution_count": null, "outputs": [], "id": "a18485d3"}, {"cell_type": "markdown", "source": "## Generate SQL for data extraction", "metadata": {}, "id": "8bb6836f"}, {"cell_type": "markdown", "source": "Alternatively, we can generate a SQL query instead of extracting the data. This SQL query can be run in a spark-enabled cluster to retrieve data. To generate a SQL query, we use the `--sql` flag and print it on STDOUT.", "metadata": {}, "id": "4b86f632"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"germline\", cohort, \"--retrieve-annotation\", \"{\\\"gene_name\\\": [\\\"TP53\\\"]}\", \"--sql\", \"-o\", \"-\"]\nprint(subprocess.check_output(cmd).decode())", "metadata": {}, "execution_count": null, "outputs": [], "id": "bd699234"}, {"cell_type": "markdown", "source": "## Return data from a specific assay", "metadata": {}, "id": "53219983"}, {"cell_type": "markdown", "source": "We provide multi-assay support, where there may be one, or many of the same assay type. If multiple assays of the same type exist in a dataset, it is possible to specify the exact assay to query from using the argument, `--assay-name`. Default functionality is to simply return data from the first eligible assay in the dataset.", "metadata": {}, "id": "2eed36a0"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"germline\", dataset, \"--assay-name\", \"UK Biobank Synthetic 100k_assay\", \"--retrieve-annotation\", \"{\\\"gene_name\\\": [\\\"TP53\\\"]}\", \"--sql\", \"-o\", \"out.sql\"]\nsubprocess.check_call(cmd)", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "99eb229a"}]}