{"metadata": {"kernelspec": {"display_name": "Python 3 (ipykernel)", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.9.16"}}, "nbformat_minor": 5, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# \u201cdx extract_assay somatic\u201d in Python\n<hr/>\n***As-Is Software Disclaimer***\n\nThis content in this repository is delivered \u201cAs-Is\u201d. Notwithstanding anything to the contrary, DNAnexus will have no warranty, support, liability or other obligations with respect to Materials provided hereunder.\n\n<hr/>\n\nThis notebook demonstrates usage of the dx command `extract_assay somatic` to:\n* Retrieve somatic variants and associated annotations from a Dataset and Cohort.\n* Compare a tumor sample to a respective normal sample from a Dataset and Cohort.\n\n<a href=\"https://github.com/dnanexus/OpenBio/blob/master/LICENSE.md\">MIT License</a> applies to this notebook.", "metadata": {}, "id": "f813a9f8-e40c-45ff-964c-af778d610790"}, {"cell_type": "markdown", "source": "## Preparing your environment\n### Launch spec:\n\n* App name: JupyterLab with Python, R, Stata, ML\n* Kernel: Python\n* Instance type: mem1_ssd1_v2_x2\n* Cost: < $0.1\n* Runtime: < 2 min\n* Data description: Input for this notebook is a v3.0 Dataset or Cohort object ID.", "metadata": {}, "id": "36ab692d-d6a8-4f2f-8944-fb624d2e789d"}, {"cell_type": "markdown", "source": "### Install DNAnexus supported package, dxpy", "metadata": {}, "id": "5f9723cb-394e-4a4f-937f-83047ece76a9"}, {"cell_type": "code", "source": "# For dx extract_assay somatic, dxpy must be v0.352.1 or greater\n# However, a more recent version of dxpy on PyPI may already be available\n# and installed, making the below \"pip\" install unecessary.\n!pip3 install -U dxpy==0.363.0", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "c225718d-1042-4738-9950-59aa31e921ff"}, {"cell_type": "code", "source": "import subprocess\nimport dxpy\nimport pandas as pd\nimport os\nimport glob\nimport json\npd.set_option('display.max_columns', None)", "metadata": {}, "execution_count": null, "outputs": [], "id": "96d3828d-fece-436b-8f98-29e3eea40f46"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"somatic\", \"--help\"]\nhelp_text = subprocess.check_output(cmd)\nprint(help_text.decode(), end=\"\\n\")", "metadata": {}, "execution_count": null, "outputs": [], "id": "cf2c5736-b905-4895-b6ed-c9bee6af5644"}, {"cell_type": "markdown", "source": "### Assign environment variables", "metadata": {}, "id": "9f7452f7-14cf-4df3-9507-c369a73f3625"}, {"cell_type": "code", "source": "# The referenced dataset is not a public dataset and listed here only to demonstrate as an example input.\n# The user will need to supply a permissible and valid project ID and record ID\n\n# Assign a project qualified dataset, project-id:record-id\ndataset=\"project-GXYZ82Q04YzF5q68Fyqp9PbX:record-GXYpp3j04YzGx2zFyk8Yykvz\"\ncohort=\"project-GXYZ82Q04YzF5q68Fyqp9PbX:record-GXqQPK804YzJ3yB9Y6B231gX\"", "metadata": {}, "execution_count": null, "outputs": [], "id": "7f5a601d-416c-436f-ae17-d84e9823a56c"}, {"cell_type": "markdown", "source": "## Extract data from a dataset", "metadata": {}, "id": "b7f561a0-74c6-4a1e-ab08-8c285308e26e"}, {"cell_type": "markdown", "source": "### 1. Explore the somatic assays of a dataset\n\nCheck which somatic variant assays are available in the dataset using the `--list-assays` flag.", "metadata": {}, "id": "0d903fbd-772e-445d-b4e3-17548a7de947"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"somatic\", dataset, \"--list-assays\"]\nlist_assay = subprocess.check_output(cmd)\nprint(list_assay.decode(), end=\"\\n\")", "metadata": {}, "execution_count": null, "outputs": [], "id": "82ad238f-4c59-46a8-ac5c-6a4513f9f1ce"}, {"cell_type": "markdown", "source": "### 2. Retrieve somatic data from the dataset\n\nData may be retrieved using the option, `--retrieve-variant`. The option accepts a JSON object as input for filtering purposes, where the object is supplied either as a \\*.json file or as a string. For additional help, use `--json-help` trailing `--retrieve-variant` for both a template and a list of filters available to the respective option.", "metadata": {}, "id": "538f8bfd-02e7-40fa-a539-48a1fbb9ba7b"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"somatic\", dataset, \"--retrieve-variant\", \"--json-help\"]\njson_help = subprocess.check_output(cmd).decode()\nprint(json_help)", "metadata": {}, "execution_count": null, "outputs": [], "id": "a6b9ad90-b164-4da4-a1e7-927f36b79ab6"}, {"cell_type": "markdown", "source": "#### Example on how to create a JSON of filters to retrieve data\n\nJSON may be created in many ways, including writing the object to a file.", "metadata": {}, "id": "367e0fea-47a2-4600-a87c-858f1dba0566"}, {"cell_type": "code", "source": "filter_dict = {\n    \"location\": [\n        {\n            \"chromosome\": \"chr1\",\n            \"starting_position\": \"10000000\",\n            \"ending_position\": \"14000000\"\n        }\n    ]\n}\nfilter = json.dumps(filter_dict)\n\nwith open(\"filter.json\", \"w\") as op:\n    op.write(filter)", "metadata": {}, "execution_count": null, "outputs": [], "id": "fc9ed026-8bf1-49ad-94c1-48a292fd5f8c"}, {"cell_type": "code", "source": "cmd = [\"cat\", \"filter.json\"]\nallele_filter_file = subprocess.check_output(cmd).decode()\nprint(allele_filter_file)", "metadata": {}, "execution_count": null, "outputs": [], "id": "a04c6993-cf12-4231-b464-e8de0838ea91"}, {"cell_type": "markdown", "source": "#### Use the generated JSON file and retrieve variant data\n\nList allele IDs and related information based on the variant filters mentioned in the variant_filter.json. The output data can be printed to STDOUT or written to a file. Here we print it to STDOUT.", "metadata": {}, "id": "83b92789-db67-4320-93be-71f6ebf6c7de"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"somatic\", dataset, \"--retrieve-variant\", \"filter.json\", \"-o\", \"-\"]\nsubprocess.check_call(cmd)", "metadata": {}, "execution_count": null, "outputs": [], "id": "c2963c6c-cc59-4722-8749-90e6c7b54c38"}, {"cell_type": "markdown", "source": "#### Add additional fields to output\n\nFor a full list of fields available for output, use the flag, `--additional-fields-help`.", "metadata": {}, "id": "abc8e90d-333f-4e6b-8b14-372b46c2de9c"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"somatic\", dataset, \"--additional-fields-help\"]\nsubprocess.check_call(cmd)", "metadata": {}, "execution_count": null, "outputs": [], "id": "16ad6e94-2e30-4f8a-8c0c-1f04a7dd7611"}, {"cell_type": "markdown", "source": "Specify additional fields using `--additional-fields`", "metadata": {}, "id": "9340571d-6cf3-4576-86ce-890c76932f83"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"somatic\", dataset,\n       \"--retrieve-variant\", \"filter.json\",\n       \"--additional-fields\", \"sample_id,tumor_normal\",\n       \"-o\", \"-\"]\nsubprocess.check_call(cmd)", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "1bc38f9e-374b-4389-8973-8d00a527411f"}, {"cell_type": "markdown", "source": "#### Additionally, list normal samples alongside tumor samples\n\nBy default, normal samples are excluded from results. To include, use the flag, `--include-normal-sample`.", "metadata": {}, "id": "258fa217-7cfd-45f3-8061-616df40e8b97"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"somatic\", dataset,\n       \"--retrieve-variant\", \"filter.json\",\n       \"--additional-fields\", \"sample_id,tumor_normal\",\n       \"--include-normal-sample\",\n       \"-o\", \"-\"]\nsubprocess.check_call(cmd)", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "e205d7ca-f05c-4d4e-a815-996aad8bd25b"}, {"cell_type": "markdown", "source": "#### Generate a list of unique sample IDs\n\nHere, we use a JSON string for filtering as opposed to a .json file. We write the retrieved data to the `retrieve_output.tsv` file using the `--output <filename>` option. Additionally, we can include information to map a tumor sample to a normal sample and VCF data such as the FORMAT and GENOTYPE fields, verbatim.", "metadata": {}, "id": "60806fe4-e605-4975-a546-7273598fc482"}, {"cell_type": "code", "source": "filter_str = '{\"location\": [{\"chromosome\": \"chr1\", \"starting_position\": \"10000000\", \"ending_position\": \"14000000\"}]}'\n\ncmd = [\"dx\", \"extract_assay\", \"somatic\", dataset,\n       \"--retrieve-variant\", filter_str,\n       \"--additional-fields\", \"sample_id,tumor_normal,normal_assay_sample_id,normal_allele_ids,FORMAT,GENOTYPE\",\n       \"--include-normal-sample\",\n       \"--output\", \"retrieve_output.tsv\"]\nsubprocess.check_call(cmd)", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "98f2e70a-22df-47fc-b339-3f0e4e454df5"}, {"cell_type": "code", "source": "retrieve_genotype_output = pd.read_csv(\"retrieve_output.tsv\", sep=\"\\t\")\nretrieve_genotype_output.head()", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "1a91beba-17c8-43ca-b435-79dd57dbd419"}, {"cell_type": "markdown", "source": "Parse the output for unique sample IDs", "metadata": {}, "id": "7b949d02-c125-49c1-9dc2-616fee9192bb"}, {"cell_type": "code", "source": "retrieve_genotype_output.sample_id.unique()", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "72133b5d-091a-4858-ab01-5195b6896bcc"}, {"cell_type": "markdown", "source": "#### Use VCF metadata to assist in parsing data\n\nVCF fields, such as INFO and FORMAT may contain multiple IDs having additional context specific to your data. These IDs and context are described in the metadata of the VCF, and available using the flag, `--retrieve-meta-info`. This context is helpful (often, necessary) for further parsing INFO and FORMAT sections of a VCF.", "metadata": {}, "id": "2b309823-1060-422b-850d-594048e94540"}, {"cell_type": "code", "source": "cmd = [\"dx\", \"extract_assay\", \"somatic\", dataset, \"--retrieve-meta-info\", \"-o\", \"-\"]\nsubprocess.check_call(cmd)", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "a65b64ce-6d6c-43d6-a332-7633cee6d0dc"}, {"cell_type": "markdown", "source": "## Extract data from a cohort\n\nHere we will retrieve all the variants in the gene `ENSG00000120942` and their associated annotations from a saved cohort.", "metadata": {}, "id": "dd9f0cdd-5fbc-46d0-a169-885605896ec8"}, {"cell_type": "markdown", "source": "#### Retrieve sample IDs based on allele IDs\n\nUse the allele IDs generated in the previous step as a filter and get sample IDs and related genotype information using the `--retrieve-genotype` filter. Here, we use a JSON string for filtering as opposed to a .json file. ", "metadata": {}, "id": "512b2645-dd26-4bd8-9e35-8161b234e26a"}, {"cell_type": "code", "source": "filter_str = '{\"annotation\": {\"gene\": [\"ENSG00000120942\"]}}'\ncmd = [\"dx\", \"extract_assay\", \"somatic\", cohort, \"--retrieve-variant\", filter_str, \"-o\", \"-\"]\nsubprocess.check_call(cmd)", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "3c821886-a5a2-4628-a4d0-621a686cfd47"}, {"cell_type": "markdown", "source": "## Generate SQL for data extraction\n\nAlternatively, we can generate a SQL query instead of extracting the data. This SQL query can be run in a spark-enabled cluster to retrieve data. To generate a SQL query, we use the `--sql` flag and print it on STDOUT.", "metadata": {}, "id": "55d97169-755e-4e07-98cb-4ab9dbae7ecd"}, {"cell_type": "code", "source": "filter_str = '{\"annotation\": {\"gene\": [\"ENSG00000120942\"]}}'\ncmd = [\"dx\", \"extract_assay\", \"somatic\", cohort, \"--retrieve-variant\", filter_str, \"--sql\", \"-o\", \"-\"]\nsubprocess.check_call(cmd)", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "82fec825-1525-4a1c-9fb8-f530b6db87c7"}, {"cell_type": "markdown", "source": "## Return data from a specific assay\n\nWe provide multi-assay support, where there may be one, or many of the same assay type. If multiple assays of the same type exist in a dataset, it is possible to specify the exact assay to query from using the argument, --assay-name. Default functionality is to simply return data from the first eligible assay in the dataset.", "metadata": {}, "id": "b0a7ca84-6f46-4147-8d44-50aeb41b7980"}, {"cell_type": "code", "source": "filter_str = '{\"annotation\": {\"gene\": [\"ENSG00000120942\"]}}'\ncmd = [\"dx\", \"extract_assay\", \"somatic\", dataset,\n       \"--assay-name\", \"example_sval_202307062200\",\n       \"--retrieve-variant\", filter_str,\n       \"--sql\", \"-o\", \"out.sql\"]\nsubprocess.check_call(cmd)", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "99e52092-f5d5-4f0b-8646-5682bca030d4"}]}