{"metadata": {"kernelspec": {"display_name": "Bash", "language": "bash", "name": "bash"}, "language_info": {"codemirror_mode": "shell", "file_extension": ".sh", "mimetype": "text/x-sh", "name": "bash"}}, "nbformat_minor": 5, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# \u201cdx extract_assay germline\u201d in Bash\n<hr/>\n***As-Is Software Disclaimer***\n\nThis content in this repository is delivered \u201cAs-Is\u201d. Notwithstanding anything to the contrary, DNAnexus will have no warranty, support, liability or other obligations with respect to Materials provided hereunder.\n\n<hr/>\n\nThis notebook demonstrates usage of the dx command `extract_assay germline` to:\n* Get a list of sample-ids based corresponding to a set of RSIDs from a dataset.\n* Retrieve all the alleles in one gene and their associated annotation from a cohort.\n\n<a href=\"https://github.com/dnanexus/OpenBio/blob/master/LICENSE.md\">MIT License</a> applies to this notebook.", "metadata": {}, "id": "37573b11"}, {"cell_type": "markdown", "source": "## Preparing your environment\n### Launch spec:\n\n* App name: JupyterLab with Python, R, Stata, ML\n* Kernel: Bash\n* Instance type: mem1_ssd1_v2_x2\n* Cost: < $0.1\n* Runtime: =~ 2 min\n* Data description: Input for this notebook is a v3.0 Dataset or Cohort object ID. In the example considered below, the dataset and cohort have 50K samples and around 16M variants.", "metadata": {}, "id": "2669791f"}, {"cell_type": "markdown", "source": "### Install DNAnexus supported package, dxpy", "metadata": {}, "id": "85b61192"}, {"cell_type": "code", "source": "# For dx extract_assay germline, dxpy must be v0.349.1 or greater\n# However, a more recent version of dxpy on PyPI may already be available\n# and installed, making the below \"pip\" install unecessary.\npip3 install -U dxpy==0.363.0", "metadata": {}, "execution_count": null, "outputs": [], "id": "6768d022"}, {"cell_type": "code", "source": "dx extract_assay germline --help", "metadata": {}, "execution_count": null, "outputs": [], "id": "97c59763"}, {"cell_type": "markdown", "source": "### Assign environment variables", "metadata": {}, "id": "748eb9dc"}, {"cell_type": "code", "source": "# The referenced dataset is not a public dataset and listed here only to demonstrate as an example input.\n# The user will need to supply a permissible and valid project ID and record ID\n\n# Assign a project qualified dataset, project-id:record-id\ndataset=\"project-GXYXjj00j49b5qX2Kzq1qbZk:record-FykXjyj0pZ7B8gKvKkxFF7QJ\"\ncohort=\"project-GXYXjj00j49b5qX2Kzq1qbZk:record-GXYkQyQ0j49jZ2fKpKZbxZ57\"", "metadata": {}, "execution_count": null, "outputs": [], "id": "bffce035"}, {"cell_type": "markdown", "source": "## Extract data from a dataset", "metadata": {}, "id": "83b7085e"}, {"cell_type": "markdown", "source": "### 1. Explore the genetic assays of a dataset\n\nCheck which genetic variant assays are available in the dataset using the `--list-assays` flag.", "metadata": {}, "id": "f882048b"}, {"cell_type": "code", "source": "dx extract_assay germline ${dataset} --list-assays", "metadata": {}, "execution_count": null, "outputs": [], "id": "13b743ab"}, {"cell_type": "markdown", "source": "### 2. Retrieve genomic data from the dataset", "metadata": {}, "id": "d8ee9c59"}, {"cell_type": "markdown", "source": "Data may be retrieved using one of three distinct options, `--retrieve-allele`, `--retrieve-annotation`, or `--retrieve-genotype`. Each option accepts a JSON object as input for filtering purposes, where the object is supplied either as a \\*.json file or as a string. For additional help, use `--json-help` trailing one of the `--retrieve-allele`, `--retrieve-annotation`, or `--retrieve-genotype` options for both a template and a list of filters available to the respective option.", "metadata": {}, "id": "0618ef2c"}, {"cell_type": "code", "source": "dx extract_assay germline ${dataset} --retrieve-allele --json-help", "metadata": {}, "execution_count": null, "outputs": [], "id": "bd286ae2"}, {"cell_type": "markdown", "source": "#### Example on how to create a JSON of filters to retrieve data\n\nA common way to create a json object used for retrieving data is to update the json template from `--json-help` with the filters needed for the situation.", "metadata": {}, "id": "3c7ebff0"}, {"cell_type": "code", "source": "dx extract_assay germline ${dataset} --retrieve-allele --json-help \\\n| grep -v '^#' \\\n> allele_filter_template.json", "metadata": {}, "execution_count": null, "outputs": [], "id": "0b8675af"}, {"cell_type": "code", "source": "cat allele_filter_template.json", "metadata": {}, "execution_count": null, "outputs": [], "id": "acb85c60"}, {"cell_type": "markdown", "source": "Change the rsID values to the desired values, select the type of allele as SNP and remove all other filters from the JSON file.", "metadata": {}, "id": "8a02cc1a"}, {"cell_type": "code", "source": "jq '{rsid, type}' allele_filter_template.json \\\n| jq '.rsid |= [\"rs1342568097\", \"rs1267100748\"]' \\\n| jq '.type |= [\"SNP\"]' \\\n> allele_filter.json", "metadata": {}, "execution_count": null, "outputs": [], "id": "dc58492c"}, {"cell_type": "code", "source": "cat allele_filter.json", "metadata": {}, "execution_count": null, "outputs": [], "id": "3158c6a6"}, {"cell_type": "markdown", "source": "Alternatively, the \\*.json file may be generated *de novo*.", "metadata": {}, "id": "f4eb40c6"}, {"cell_type": "markdown", "source": "#### Use the generated JSON file and retrieve allele data\n\nList allele IDs and related information based on the allele filters mentioned in the allele_filter.json. The output data can be printed to STDOUT or written to a .tsv file. Here we print it to STDOUT.", "metadata": {}, "id": "9865457e"}, {"cell_type": "code", "source": "dx extract_assay germline ${dataset} \\\n--retrieve-allele allele_filter.json \\\n-o -", "metadata": {}, "execution_count": null, "outputs": [], "id": "1d8a1bcd"}, {"cell_type": "markdown", "source": "#### Retrieve sample IDs based on allele IDs\n\nUse the allele IDs generated in the previous step as a filter and get sample IDs and related genotype information using the `--retrieve-genotype` filter. Here, we use a JSON string for filtering as opposed to a \\*.json file. We write the retrieved data to the `retrieve_genotype_output.tsv` file using the `--output OUTPUT` option.", "metadata": {}, "id": "0ea2941d"}, {"cell_type": "code", "source": "dx extract_assay germline ${dataset} \\\n--retrieve-genotype '{\"allele_id\": [\"18_47359_G_T\", \"18_47360_C_T\"]}' \\\n--output 'retrieve_genotype_output.tsv'", "metadata": {}, "execution_count": null, "outputs": [], "id": "75e964d1"}, {"cell_type": "code", "source": "head retrieve_genotype_output.tsv", "metadata": {}, "execution_count": null, "outputs": [], "id": "8d7eac70"}, {"cell_type": "markdown", "source": "Get a list of sample IDs", "metadata": {}, "id": "ef820618"}, {"cell_type": "code", "source": "cut -f1 retrieve_genotype_output.tsv", "metadata": {}, "execution_count": null, "outputs": [], "id": "3598bde5"}, {"cell_type": "markdown", "source": "## Extract data from a cohort\n\nHere we will retrieve all the alleles in the gene, TP53, and the associated annotations from a saved cohort.", "metadata": {}, "id": "5a10e181"}, {"cell_type": "markdown", "source": "#### Check the available filters in the `--retrieve-annotation` option.", "metadata": {}, "id": "71beae3c"}, {"cell_type": "code", "source": "dx extract_assay germline ${cohort} --retrieve-annotation --json-help", "metadata": {}, "execution_count": null, "outputs": [], "id": "94e860f6"}, {"cell_type": "markdown", "source": "#### We will use the `gene_name` filter and save the output data to a .tsv file.", "metadata": {}, "id": "3913a069"}, {"cell_type": "code", "source": "dx extract_assay germline ${cohort} \\\n--retrieve-annotation '{\"gene_name\": [\"TP53\"]}' \\\n--output \"TP53_allele_annotation.tsv\"", "metadata": {}, "execution_count": null, "outputs": [], "id": "6caed237"}, {"cell_type": "code", "source": "head TP53_allele_annotation.tsv", "metadata": {}, "execution_count": null, "outputs": [], "id": "00326214"}, {"cell_type": "markdown", "source": "## Generate SQL for data extraction", "metadata": {}, "id": "f0a82bca"}, {"cell_type": "markdown", "source": "Alternatively, we can generate a SQL query instead of extracting the data. This SQL query can be run in a spark-enabled cluster to retrieve data. To generate a SQL query, we use the `--sql` flag and print it on STDOUT.", "metadata": {}, "id": "c256b4a7"}, {"cell_type": "code", "source": "dx extract_assay germline ${cohort} \\\n--retrieve-annotation '{\"gene_name\": [\"TP53\"]}' \\\n--sql \\\n-o -", "metadata": {}, "execution_count": null, "outputs": [], "id": "463fc804"}, {"cell_type": "markdown", "source": "## Return data from a specific assay", "metadata": {}, "id": "d9a4c92f"}, {"cell_type": "markdown", "source": "We provide multi-assay support, where there may be one, or many of the same assay type. If multiple assays of the same type exist in a dataset, it is possible to specify the exact assay to query from using the argument, `--assay-name`. Default functionality is to simply return data from the first eligible assay in the dataset.", "metadata": {}, "id": "e7954749"}, {"cell_type": "code", "source": "dx extract_assay germline ${dataset} \\\n--assay-name 'UK Biobank Synthetic 100k_assay' \\\n--retrieve-annotation '{\"gene_name\": [\"TP53\"]}' \\\n--sql \\\n-o 'out.sql'", "metadata": {"tags": []}, "execution_count": null, "outputs": [], "id": "f074220c"}]}