{"metadata": {"kernelspec": {"name": "ir", "display_name": "R", "language": "R"}, "language_info": {"name": "R", "codemirror_mode": "r", "pygments_lexer": "r", "mimetype": "text/x-r-source", "file_extension": ".r", "version": "4.2.0"}}, "nbformat_minor": 5, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# \u201cdx extract_assay germline\u201d in R\n<hr/>\n***As-Is Software Disclaimer***\n\nThis content in this repository is delivered \u201cAs-Is\u201d. Notwithstanding anything to the contrary, DNAnexus will have no warranty, support, liability or other obligations with respect to Materials provided hereunder.\n\n<hr/>\n\nThis notebook demonstrates usage of the dx command `extract_assay germline` to:\n* Get a list of sample-ids based corresponding to a set of RSIDs from a dataset.\n* Retrieve all the alleles in one gene and their associated annotation from a cohort.\n\n<a href=\"https://github.com/dnanexus/OpenBio/blob/master/LICENSE.md\">MIT License</a> applies to this notebook.", "metadata": {}}, {"cell_type": "markdown", "source": "## Preparing your environment\n### Launch spec:\n\n* App name: JupyterLab with Python, R, Stata, ML, Image Processing\n* Kernel: R\n* Instance type: mem1_ssd1_v2_x2\n* Cost: < $0.2\n* Runtime: ~= 5 min\n* Data description: Input for this notebook is a v3.0 Dataset or Cohort object ID. In the example considered below, the dataset and cohort have 50K samples and around 16M variants.", "metadata": {}}, {"cell_type": "markdown", "source": "### Install DNAnexus supported package, dxpy", "metadata": {}}, {"cell_type": "code", "source": "# For dx extract_assay germline, dxpy must be v0.349.1 or greater\n# However, a more recent version of dxpy on PyPI may already be available\n# and installed, making the below \"pip\" install unecessary.\nsystem(\"pip3 install -U dxpy==0.350.1\")", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "install.packages(\"readr\")", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "library(dplyr)\nlibrary(readr)\nlibrary(stringr)\nlibrary(jsonlite)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "cmd <- \"dx extract_assay germline --help\"\nsystem(cmd, intern = TRUE)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "### Assign environment variables", "metadata": {}}, {"cell_type": "code", "source": "# The referenced dataset is not a public dataset and listed here only to demonstrate as an example input.\n# The user will need to supply a permissible and valid project ID and record ID\n\n# Assign a project qualified dataset, project-id:record-id\ndataset=\"project-GXYXjj00j49b5qX2Kzq1qbZk:record-FykXjyj0pZ7B8gKvKkxFF7QJ\"\ncohort=\"project-GXYXjj00j49b5qX2Kzq1qbZk:record-GXYkQyQ0j49jZ2fKpKZbxZ57\"", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "## Extract data from a dataset", "metadata": {}}, {"cell_type": "markdown", "source": "### 1. Explore the genetic assays of a dataset\n\nCheck which genetic variant assays are available in the dataset using the `--list-assays` flag.", "metadata": {}}, {"cell_type": "code", "source": "cmd <- paste(\"dx extract_assay germline\", dataset, \"--list-assays\")\nprint(system(cmd, intern = TRUE))", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "### 2. Retrieve genomic data from the dataset", "metadata": {}}, {"cell_type": "markdown", "source": "Data may be retrieved using one of three distinct options, `--retrieve-allele`, `--retrieve-annotation`, or `--retrieve-genotype`. Each option accepts a JSON object as input for filtering purposes, where the object is supplied either as a \\*.json file or as a string. For additional help, use `--json-help` trailing one of the `--retrieve-allele`, `--retrieve-annotation`, or `--retrieve-genotype` options for both a template and a list of filters available to the respective option.", "metadata": {}}, {"cell_type": "code", "source": "cmd <- paste(\"dx extract_assay germline\", dataset, \"--retrieve-allele --json-help\")\nprint(system(cmd, intern = TRUE), quote=FALSE)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "#### Example on how to create a JSON of filters to retrieve data\n\nA common way to create a json object used for retrieving data is to update the json template from `--json-help` with the filters needed for the situation.", "metadata": {}}, {"cell_type": "code", "source": "allele_filter <- list(rsid = c(\"rs1342568097\", \"rs1267100748\"), type = c(\"SNP\"))\n\nwrite_json(allele_filter, \"allele_filter.json\")", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "#### Use the generated JSON file and retrieve allele data\n\nList allele IDs and related information based on the allele filters mentioned in the allele_filter.json and save the results in the `retrieve_allele_output.tsv` file using the `--output <filename>` option.", "metadata": {}}, {"cell_type": "code", "source": "cmd <- paste(\"dx extract_assay germline\", dataset, \"--retrieve-allele allele_filter.json -o retrieve_allele_output.tsv\")\nsystem(cmd)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "retrieve_allele_output <- readr::read_tsv(\"retrieve_allele_output.tsv\", show_col_types = FALSE)\nhead(retrieve_allele_output)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "#### Retrieve sample IDs based on allele IDs\n\nUse the allele IDs generated in the previous step as a filter and get sample IDs and related genotype information using the `--retrieve-genotype` filter. Here, we use a JSON string for filtering as opposed to a \\*.json file.", "metadata": {}}, {"cell_type": "code", "source": "# Create filter\ngenotype_filter <- toJSON(list(allele_id = retrieve_allele_output$allele_id))\n\n# Run extract_assay command\ncmd <- paste(\"dx extract_assay germline\", dataset, \"--retrieve-genotype '\", as.character(genotype_filter), \"' -o retrieve_genotype_output.tsv\")\nsystem(cmd)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "retrieve_genotype_output <- readr::read_tsv(\"retrieve_genotype_output.tsv\", show_col_types = FALSE)\nhead(retrieve_genotype_output)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Get a list of sample IDs", "metadata": {}}, {"cell_type": "code", "source": "print(\"List of sample IDs with previously derived allele IDs.\")\nretrieve_genotype_output$sample_id", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "## Extract data from a cohort\n\nHere we will retrieve all the alleles in the gene `TP53` and their associated annotations from a saved cohort.", "metadata": {}}, {"cell_type": "markdown", "source": "#### Check the available filters in the `--retrieve-annotation` option.", "metadata": {}}, {"cell_type": "code", "source": "cmd <- paste(\"dx extract_assay germline\", cohort, \"--retrieve-annotation --json-help\")\nprint(system(cmd, intern=TRUE), quote = FALSE)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "We will use the `gene_name` filter and save the output data to a .tsv file.", "metadata": {}}, {"cell_type": "code", "source": "cmd <- paste(\"dx extract_assay germline\", cohort, \"--retrieve-annotation '{\\\"gene_name\\\": [\\\"TP53\\\"]}' -o TP53_allele_annotation.tsv\")\nsystem(cmd)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "TP53_allele_annotation <- readr::read_tsv(\"TP53_allele_annotation.tsv\", show_col_types = FALSE)\nhead(TP53_allele_annotation)", "metadata": {"tags": [], "trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "## Generate SQL for data extraction", "metadata": {}}, {"cell_type": "markdown", "source": "Alternatively, we can generate a SQL query instead of extracting the data. This SQL query can be run in a spark-enabled cluster to retrieve data. To generate a SQL query, we use the `--sql` flag and print it on STDOUT.", "metadata": {}}, {"cell_type": "code", "source": "cmd <- paste(\"dx extract_assay germline\", cohort, \"--retrieve-annotation '{\\\"gene_name\\\": [\\\"TP53\\\"]}' --sql -o -\")\nprint(system(cmd, intern = TRUE), quote=FALSE)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "## Return data from a specific assay", "metadata": {}}, {"cell_type": "markdown", "source": "We provide multi-assay support, where there may be one, or many of the same assay type. If multiple assays of the same type exist in a dataset, it is possible to specify the exact assay to query from using the argument, `--assay-name`. Default functionality is to simply return data from the first eligible assay in the dataset.", "metadata": {"tags": []}}, {"cell_type": "code", "source": "cmd <- paste(\"dx extract_assay germline\", dataset, \"--assay-name 'UK Biobank Synthetic 100k_assay' --retrieve-annotation '{\\\"gene_name\\\": [\\\"TP53\\\"]}' --sql -o out.sql\")\nsystem(cmd)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}]}